### ISSUE TITLE
Unable to build with CMake(Compile error by macro)

### ISSUE THREAD

**Issue by Remering** [vurtun/nuklear#770]
Here's the CMakeList.txt
```
cmake_minimum_required(VERSION 3.12)
project(StudentManagementSystem C)

include_directories(D:\\WorkPlace\\C\\nuklear)
include_directories(D:\\WorkPlace\\C\\nuklear\\src)
set(CMAKE_C_STANDARD 99)

add_executable(StudentManagementSystem main.c)
```

My code:
```
#include <stdio.h>
#include <nuklear.h>
#include <stdlib.h>

#define NK_IMPLEMENTATION
#define XMS 512 //unit: kb


enum {EASY, HARD};
int main() {
    static int op = EASY;
    static float value = 0.6f;
    static int i = 20;
    struct nk_context ctx;
    void *mem_holder = malloc(XMS * 1024 * 8);
    if (!mem_holder){
        puts("Error, unable to allocate enough memory for me.");
        printf("%iKB needed", XMS);
        return 1;
    }
    nk_init_fixed(&ctx, mem_holder, XMS, NULL);
    if (nk_begin(&ctx, "Show", nk_rect(50, 50, 220, 220),
            NK_WINDOW_BORDER|NK_WINDOW_MOVABLE|NK_WINDOW_CLOSABLE)){
        nk_layout_row_static(&ctx, 30, 80, 1);
        if (nk_button_label(&ctx, "button")){
            puts("Button button clicked");
        }
        nk_layout_row_dynamic(&ctx, 30, 2);
        if (nk_option_label(&ctx, "easy", op == EASY)){
            op = EASY;
            puts("Now op turns easy");
        }
        if (nk_option_label(&ctx, "hard", op == HARD)){
            op = HARD;
            puts("Now op turns hard");
        }
        nk_layout_row_begin(&ctx, NK_STATIC, 30, 2);
        {
            nk_layout_row_push(&ctx, 50);
            nk_label(&ctx, "Volume:", NK_TEXT_LEFT);
            nk_layout_row_push(&ctx, 110);
            nk_slider_float(&ctx, 0, &value, 1.f, 1.f);
        }
        nk_layout_row_end(&ctx);
    }
    nk_end(&ctx);
}
```

Build message: 
```
C:\Users\dell\AppData\Local\JetBrains\Toolbox\apps\CLion\ch-0\182.5107.21\bin\cmake\win\bin\cmake.exe --build D:\WorkPlace\C\StudentManagementSystem\cmake-build-debug --target StudentManagementSystem -- -j 6
Scanning dependencies of target StudentManagementSystem
[ 50%] Building C object CMakeFiles/StudentManagementSystem.dir/main.c.obj
In file included from D:\WorkPlace\C\StudentManagementSystem\main.c:2:
D:/WorkPlace/C/nuklear/nuklear.h:291:61: error: size of array '_dummy_array422' is negative
   #define NK_STATIC_ASSERT(exp) typedef char NK_UNIQUE_NAME(_dummy_array)[(exp)?1:-1]
                                                             ^~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:280:46: note: in definition of macro 'NK_STRING_JOIN_IMMEDIATE'
 #define NK_STRING_JOIN_IMMEDIATE(arg1, arg2) arg1 ## arg2
                                              ^~~~
D:/WorkPlace/C/nuklear/nuklear.h:282:36: note: in expansion of macro 'NK_STRING_JOIN_DELAY'
 #define NK_STRING_JOIN(arg1, arg2) NK_STRING_JOIN_DELAY(arg1, arg2)
                                    ^~~~~~~~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:287:32: note: in expansion of macro 'NK_STRING_JOIN'
   #define NK_UNIQUE_NAME(name) NK_STRING_JOIN(name,__LINE__)
                                ^~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:291:46: note: in expansion of macro 'NK_UNIQUE_NAME'
   #define NK_STATIC_ASSERT(exp) typedef char NK_UNIQUE_NAME(_dummy_array)[(exp)?1:-1]
                                              ^~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:422:1: note: in expansion of macro 'NK_STATIC_ASSERT'
 NK_STATIC_ASSERT(sizeof(nk_size) >= sizeof(void*));
 ^~~~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:291:61: error: size of array '_dummy_array423' is negative
   #define NK_STATIC_ASSERT(exp) typedef char NK_UNIQUE_NAME(_dummy_array)[(exp)?1:-1]
                                                             ^~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:280:46: note: in definition of macro 'NK_STRING_JOIN_IMMEDIATE'
 #define NK_STRING_JOIN_IMMEDIATE(arg1, arg2) arg1 ## arg2
                                              ^~~~
D:/WorkPlace/C/nuklear/nuklear.h:282:36: note: in expansion of macro 'NK_STRING_JOIN_DELAY'
 #define NK_STRING_JOIN(arg1, arg2) NK_STRING_JOIN_DELAY(arg1, arg2)
                                    ^~~~~~~~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:287:32: note: in expansion of macro 'NK_STRING_JOIN'
   #define NK_UNIQUE_NAME(name) NK_STRING_JOIN(name,__LINE__)
                                ^~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:291:46: note: in expansion of macro 'NK_UNIQUE_NAME'
   #define NK_STATIC_ASSERT(exp) typedef char NK_UNIQUE_NAME(_dummy_array)[(exp)?1:-1]
                                              ^~~~~~~~~~~~~~
D:/WorkPlace/C/nuklear/nuklear.h:423:1: note: in expansion of macro 'NK_STATIC_ASSERT'
 NK_STATIC_ASSERT(sizeof(nk_ptr) >= sizeof(void*));
 ^~~~~~~~~~~~~~~~
mingw32-make.exe[3]: *** [CMakeFiles\StudentManagementSystem.dir\build.make:63: CMakeFiles/StudentManagementSystem.dir/main.c.obj] Error 1
mingw32-make.exe[2]: *** [CMakeFiles\Makefile2:72: CMakeFiles/StudentManagementSystem.dir/all] Error 2
mingw32-make.exe[1]: *** [CMakeFiles\Makefile2:84: CMakeFiles/StudentManagementSystem.dir/rule] Error 2
mingw32-make.exe: *** [Makefile:117: StudentManagementSystem] Error 2

```
IDE: CLion 2018.2.6
System: Windows 10(64Bits)
In addition, I'm sorry that I'm not enough wise to understand documents and demos included in repo.I put the example code into main because I don't know where to place it(or them? My English is broken).
Thanks for you any resonable suggestions and answers

**Comment by nmielcarek** [comment]
First step, move the #define NK_IMPLEMENTATION line before #include <nuklear.h>

**Comment by Gnimuc** [comment]
same here, compile failed in a cross-compile environment, my code only contains:
```
#define NK_IMPLEMENTATION
#include "nuklear.h"
```

**Comment by Gnimuc** [comment]
strange, it looks like nuklear only fails in x86_64_mingw32, but it works fine with this configuration:
```julia
#define NK_IMPLEMENTATION
#define NK_INCLUDE_FIXED_TYPES
#define NK_INCLUDE_DEFAULT_ALLOCATOR
#define NK_INCLUDE_STANDARD_IO
#define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
#define NK_INCLUDE_FONT_BAKING
#define NK_INCLUDE_DEFAULT_FONT
#include "nuklear.h"
```

**Comment by computerquip** [comment]
The static assertion at 423 (`NK_STATIC_ASSERT(sizeof(nk_ptr) >= sizeof(void*));`) is failing. 

Looking at what the definition of nk_ptr is, it appears to be a typedef of `NK_POINTER_TYPE` whose definition is determined by this large block of conditionals: 
```
#ifdef NK_INCLUDE_FIXED_TYPES
 #include <stdint.h>
 #define NK_INT8 int8_t
 #define NK_UINT8 uint8_t
 #define NK_INT16 int16_t
 #define NK_UINT16 uint16_t
 #define NK_INT32 int32_t
 #define NK_UINT32 uint32_t
 #define NK_SIZE_TYPE uintptr_t
 #define NK_POINTER_TYPE uintptr_t
#else
  #ifndef NK_INT8
    #define NK_INT8 signed char
  #endif
  #ifndef NK_UINT8
    #define NK_UINT8 unsigned char
  #endif
  #ifndef NK_INT16
    #define NK_INT16 signed short
  #endif
  #ifndef NK_UINT16
    #define NK_UINT16 unsigned short
  #endif
  #ifndef NK_INT32
    #if defined(_MSC_VER)
      #define NK_INT32 __int32
    #else
      #define NK_INT32 signed int
    #endif
  #endif
  #ifndef NK_UINT32
    #if defined(_MSC_VER)
      #define NK_UINT32 unsigned __int32
    #else
      #define NK_UINT32 unsigned int
    #endif
  #endif
  #ifndef NK_SIZE_TYPE
    #if defined(_WIN64) && defined(_MSC_VER)
      #define NK_SIZE_TYPE unsigned __int64
    #elif (defined(_WIN32) || defined(WIN32)) && defined(_MSC_VER)
      #define NK_SIZE_TYPE unsigned __int32
    #elif defined(__GNUC__) || defined(__clang__)
      #if defined(__x86_64__) || defined(__ppc64__)
        #define NK_SIZE_TYPE unsigned long
      #else
        #define NK_SIZE_TYPE unsigned int
      #endif
    #else
      #define NK_SIZE_TYPE unsigned long
    #endif
  #endif
  #ifndef NK_POINTER_TYPE
    #if defined(_WIN64) && defined(_MSC_VER)
      #define NK_POINTER_TYPE unsigned __int64
    #elif (defined(_WIN32) || defined(WIN32)) && defined(_MSC_VER)
      #define NK_POINTER_TYPE unsigned __int32
    #elif defined(__GNUC__) || defined(__clang__)
      #if defined(__x86_64__) || defined(__ppc64__)
        #define NK_POINTER_TYPE unsigned long
      #else
        #define NK_POINTER_TYPE unsigned int
      #endif
    #else
      #define NK_POINTER_TYPE unsigned long
    #endif
  #endif
#endif
```

In MinGW provided via MSYS2, both clang and gcc report that unsigned long is in fact 4 bytes. How to do this correctly can be a bit tricky if you don't want to include stdint.h or make it optional. GCC and Clang do support a built-in called `__UINTPTR_TYPE__ ` which you can be used to define your own uintptr_t alternative. However, gcc recommends against this and says you should simply include the standard headers instead and use the official types that they correspond to. More information on this page: https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html

TLDR: NK_INCLUDE_FIXED_TYPES will fix this for MinGW for now since it's an issue with the assumption of integer width.

**Comment by JamesTheHacker** [comment]
> strange, it looks like nuklear only fails in x86_64_mingw32, but it works fine with this configuration:
> 
> ```c
> #define NK_IMPLEMENTATION
> #define NK_INCLUDE_FIXED_TYPES
> #define NK_INCLUDE_DEFAULT_ALLOCATOR
> #define NK_INCLUDE_STANDARD_IO
> #define NK_INCLUDE_VERTEX_BUFFER_OUTPUT
> #define NK_INCLUDE_FONT_BAKING
> #define NK_INCLUDE_DEFAULT_FONT
> #include "nuklear.h"
> ```

I think this should be in the docs. Failed for me until I included those definitions.

**Comment by dumblob** [comment]
@JamesTheHacker yeah, it could/shall be in the docs - just issue a pull request and I'll merge it.
